### DISCUSSION TITLE
Higher-order substitution/rewrite in terms.

### DISCUSSION THREAD

**Discussion by jparsert** [cvc5/cvc5#9329] [Q&A] [answered]
Is there a higher-order version of substitute function in cvc5.
I have found the exact same question being asked before on SO in z3 [here](https://stackoverflow.com/questions/15236450/substituting-function-symbols-in-z3-formulas) and answered in detail with code.
It seems this issue has now been moved into the python API of z3.

The problem is the following say I have a conjunction `f(1,4) == 5 and f(10,20) > 30` where `f` is not defined. Now I want to substitute `f` with the term `arg_1 + arg_2`. The way I understand, the cvc5 substitute function now replaces every occurrence of `f` with `arg_1+arg_2` without also substituting the arguments `1,4` and `10,20` for `arg_1` and `arg_2` respectively.

**Comment by ajreynol** [comment, marked as answer]
There is only one version of substitution, which accounts for function substitutions. It does not automatically do beta reductions.

In detail, first note that substitutions must be well-typed. Assuming that `f` has type `Int x Int -> Int`, then the substitution in question would be `f -> (LAMBDA (VARIABLE_LIST arg_1 arg_2) (PLUS arg_1 arg_2))`, where arg_1 and arg_2 have kind `VARIABLE`.

Now, notice that beta-reduction will not happen automatically.  Applying this substitution to:
`(APPLY_UF f 1 4)`
will result in the term:
`(APPLY_UF (LAMBDA (VARIABLE_LIST arg_1 arg_2) (PLUS arg_1 arg_2)) 1 4)`

To apply beta-reduction on this term, you can use `Solver::simplify(...)`, which notice will also normalize the result, i.e. it will return `5`. Does this suffice for your use case?
